### PR TITLE
[Merged by Bors] - feat(CategoryTheory): localization of pretriangulated categories

### PR DESCRIPTION
**Pull request by joelriou**
In the new file `CategoryTheory.Localization.Triangulated`, it is shown that if a class of morphisms `W` in a pretriangulated category is compatible with the triangulation and has a left calculus of fractions, then the localized category is also pretriangulated.

---

- [x] depends on: #11737
- [x] depends on: #11721
- [x] depends on: #11728

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Localization/Triangulated.lean [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+  distinguished_cocone_triangle f := distinguished_cocone_triangle L W f
+  rotate_distinguished_triangle := L.rotate_essImageDistTriang
+  complete_distinguished_triangle_morphism := complete_distinguished_triangle_morphism L W
+
+lemma isTriangulated_functor :
+    letI : Pretriangulated D := pretriangulated L W; L.IsTriangulated :=
+    letI : Pretriangulated D := pretriangulated L W; ⟨fun T hT => ⟨T, Iso.refl _, hT⟩⟩
+
+lemma essSurj_mapArrow : L.mapArrow.EssSurj  :=
+  essSurj_mapArrow_of_hasLeftCalculusofFractions L W
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Localization/Triangulated.lean:198]: Why is this restated?

Comment (joelriou) [reply Mathlib/CategoryTheory/Localization/Triangulated.lean:198]: I found it convenient to restate it in the `Triangulated.Localization` namespace.

File: Mathlib/CategoryTheory/Localization/Triangulated.lean [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+def pretriangulated : Pretriangulated D where
+  distinguishedTriangles := L.essImageDistTriang
+  isomorphic_distinguished _ hT₁ _ e := L.essImageDistTriang_mem_of_iso e hT₁
+  contractible_distinguished :=
+    have := essSurj L W; L.contractible_mem_essImageDistTriang
+  distinguished_cocone_triangle f := distinguished_cocone_triangle L W f
+  rotate_distinguished_triangle := L.rotate_essImageDistTriang
+  complete_distinguished_triangle_morphism := complete_distinguished_triangle_morphism L W
+
+lemma isTriangulated_functor :
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/CategoryTheory/Localization/Triangulated.lean]: Thanks for the suggestion.

File: Mathlib/CategoryTheory/Localization/Triangulated.lean (hunk 1 above)
Comment (erdOne) [reply Mathlib/CategoryTheory/Localization/Triangulated.lean:198]: In that case, maybe put the `alias` below the original lemma or just rename it altogether?

Comment (joelriou) [reply Mathlib/CategoryTheory/Localization/Triangulated.lean]: I have renamed ` essSurj_mapArrow_of_hasLeftCalculusOfFractions` as `essSurj_mapArrow` (and kept `essSurj_mapArrow_of_hasRightCalculusofFractions`).

File: Mathlib/CategoryTheory/Localization/Triangulated.lean [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+lemma essImageDistTriang_mem_of_iso {T₁ T₂ : Triangle D} (e : T₂ ≅ T₁)
+    (h : T₁ ∈ L.essImageDistTriang) : T₂ ∈ L.essImageDistTriang := by
+  obtain ⟨T', e', hT'⟩ := h
+  exact ⟨T', e ≪≫ e', hT'⟩
+
+lemma contractible_mem_essImageDistTriang [EssSurj L] [HasZeroObject D]
+    [HasZeroMorphisms D] [L.PreservesZeroMorphisms] (X : D) :
+    contractibleTriangle X ∈ L.essImageDistTriang := by
+  refine' ⟨contractibleTriangle (L.objPreimage X), _, contractible_distinguished _⟩
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Localization/Triangulated.lean:66]:
```suggestion
  refine ⟨contractibleTriangle (L.objPreimage X), ?_, contractible_distinguished _⟩
```

File: Mathlib/CategoryTheory/Localization/Triangulated.lean [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  dsimp at comm₁ comm₁' comm₂ comm₂' comm₃ comm₃'
+  simp only [assoc] at comm₃
+  obtain ⟨φ, hφ₁, hφ₂⟩ := H T₁' T₂' hT₁' hT₂' (e₁.inv.hom₁ ≫ a ≫ e₂.hom.hom₁)
+    (e₁.inv.hom₂ ≫ b ≫ e₂.hom.hom₂)
+    (by simp only [assoc, ← comm₁', ← reassoc_of% fac, ← reassoc_of% comm₁])
+  have h₂ := φ.comm₂
+  have h₃ := φ.comm₃
+  dsimp at h₂ h₃
+  simp only [assoc] at h₃
+  refine' ⟨e₁.hom.hom₃ ≫ φ.hom₃ ≫ e₂.inv.hom₃, _, _⟩
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Localization/Triangulated.lean:107]:
```suggestion
  refine ⟨e₁.hom.hom₃ ≫ φ.hom₃ ≫ e₂.inv.hom₃, ?_, ?_⟩
```

File: Mathlib/CategoryTheory/Localization/Triangulated.lean [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    ∃ (Z : D) (g : Y ⟶ Z) (h : Z ⟶ X⟦(1 : ℤ)⟧),
+      Triangle.mk f g h ∈ L.essImageDistTriang := by
+  have := essSurj_mapArrow L W
+  obtain ⟨φ, ⟨e⟩⟩ : ∃ (φ : Arrow C), Nonempty (L.mapArrow.obj φ ≅ Arrow.mk f) :=
+    ⟨_, ⟨Functor.objObjPreimageIso _ _⟩⟩
+  obtain ⟨Z, g, h, H⟩ := Pretriangulated.distinguished_cocone_triangle φ.hom
+  refine' ⟨L.obj Z, e.inv.right ≫ L.map g,
+    L.map h ≫ (L.commShiftIso (1 : ℤ)).hom.app _ ≫ e.hom.left⟦(1 : ℤ)⟧', _, _, H⟩
+  refine' Triangle.isoMk _ _ (Arrow.leftFunc.mapIso e.symm) (Arrow.rightFunc.mapIso e.symm)
+    (Iso.refl _) e.inv.w.symm (by simp) _
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Localization/Triangulated.lean:137]:
```suggestion
  refine ⟨L.obj Z, e.inv.right ≫ L.map g,
    L.map h ≫ (L.commShiftIso (1 : ℤ)).hom.app _ ≫ e.hom.left⟦(1 : ℤ)⟧', _, ?_, H⟩
  refine Triangle.isoMk _ _ (Arrow.leftFunc.mapIso e.symm) (Arrow.rightFunc.mapIso e.symm)
    (Iso.refl _) e.inv.w.symm (by simp) ?_
```
I might be wrong here.

File: Mathlib/CategoryTheory/Localization/Triangulated.lean [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+  change T₃ ∈ distTriang C at hT₃
+  have hβγσ : W (β.s ≫ γ.s ≫ σ) := W.comp_mem _ _ β.hs (W.comp_mem _ _ γ.hs hσ)
+  obtain ⟨ψ₃, hψ₃, hψ₁, hψ₂⟩ := MorphismProperty.compatible_with_triangulation
+    T₂ T₃ hT₂ hT₃ α.s (β.s ≫ γ.s ≫ σ) α.hs hβγσ (by dsimp [T₃]; rw [reassoc_of% hβ])
+  let ψ : T₂ ⟶ T₃ := Triangle.homMk _ _ α.s (β.s ≫ γ.s ≫ σ) ψ₃
+    (by dsimp [T₃]; rw [reassoc_of% hβ]) hψ₁ hψ₂
+  have : IsIso (L.mapTriangle.map ψ) := Triangle.isIso_of_isIsos _
+    (inverts L W α.s α.hs) (inverts L W _ hβγσ) (inverts L W ψ₃ hψ₃)
+  refine' ⟨L.mapTriangle.map (completeDistinguishedTriangleMorphism T₁ T₃ hT₁ hT₃ α.f
+      (γ.f ≫ σ) fac.symm) ≫ inv (L.mapTriangle.map ψ), _, _⟩
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Localization/Triangulated.lean:174]:
```suggestion
  refine ⟨L.mapTriangle.map (completeDistinguishedTriangleMorphism T₁ T₃ hT₁ hT₃ α.f
      (γ.f ≫ σ) fac.symm) ≫ inv (L.mapTriangle.map ψ), ?_, ?_⟩
```

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**